x/logic/predicate/bank.go

```diff
@@ -0,0 +1,75 @@
+package predicate
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ichiban/prolog/engine"
+	"github.com/okp4/okp4d/x/logic/types"
+	"github.com/okp4/okp4d/x/logic/util"
+)
+
+// BankBalances is a predicate which unifies the given terms with the list of balances (coins) of the given account.
+//
+//	bank_balances(?Account, ?Balances)
+//
+// where:
+//   - Account represents the account address (in Bech32 format).
+//   - Coins represents the balances of the account as a list of pairs of coin denomination and amount.
+//
+// Example:
+//
+//	# Query the balances of the account.
+//	- bank_balances('okp41ffd5wx65l407yvm478cxzlgygw07h79sq0m3fm', X).
+//
+// # Query the balances of all accounts. The result is a list of pairs of account address and balances.
+// - bank_balances(X, Y).
+//
+// # Query the first balance of the given account by unifying the denomination and amount with the given terms.
+// - bank_balances('okp41ffd5wx65l407yvm478cxzlgygw07h79sq0m3fm', [-(D, A), _]).
+func BankBalances(vm *engine.VM, account, balances engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
+	return engine.Delay(func(ctx context.Context) *engine.Promise {
+		sdkContext, err := util.UnwrapSDKContext(ctx)
+		if err != nil {
+			return engine.Error(err)
+		}
+		bankKeeper := sdkContext.Value(types.BankKeeperContextKey).(types.BankKeeper)
+
+		bech32Addr := sdk.AccAddress(nil)
+		switch acc := env.Resolve(account).(type) {
+		case engine.Variable:
+		case engine.Atom:
+			bech32Addr, err = sdk.AccAddressFromBech32(acc.String())
+			if err != nil {
+				return engine.Error(fmt.Errorf("bank_spendable_coins/2: %w", err))
```

Comment:
The error naming seems to not fit regarding the predicate name. There are other similar issues i didn't point out.

```suggestion
				return engine.Error(fmt.Errorf("bank_balances/2: %w", err))
```

Reply:
Good catch! Missed that one. :+1: